src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs

```diff
@@ -394,5 +679,56 @@ internal static void SetCertificate(SafeSslHandle sslContext, SslStreamCertifica
 
             Interop.AppleCrypto.SslSetCertificate(sslContext, ptrs);
         }
+
+        public unsafe Task<SecurityStatusPalErrorCode> StartDecrypt()
+        {
+            //Debug.Assert(size > 0);
+
+            if (Tcs == null)
+            {
+                Tcs = new TaskCompletionSource<SecurityStatusPalErrorCode>();
+                Interop.AppleCrypto.NwReadFromConnection(SslContext, GCHandle.ToIntPtr(gcHandle));
+            }
+
+            return Tcs.Task;
+        }
+
+        public unsafe SecurityStatusPal PerformNwHandshake(ReadOnlySpan<byte> inputBuffer)
+        {
+            ObjectDisposedException.ThrowIf(_disposed, this);
+            if (handshakeStarted && inputBuffer.Length == 0)
+            {
+                // We may be asked to generate Alter tokens and that is not supported on macOS
+                return new SecurityStatusPal(SecurityStatusPalErrorCode.OK);
+            }
+
+            if (_handshakeDone)
+            {
+                return new SecurityStatusPal(SecurityStatusPalErrorCode.ContinueNeeded);
+            }
+            if (_framer != IntPtr.Zero && inputBuffer.Length > 0)
+            {
+                ObjectDisposedException.ThrowIf(_disposed, this);
+                fixed (byte* ptr = &MemoryMarshal.GetReference(inputBuffer))
+                {
+                    Interop.AppleCrypto.NwProcessInputData(SslContext, _framer, ptr, inputBuffer.Length);
+                }
+
+                return new SecurityStatusPal(SecurityStatusPalErrorCode.ContinuePendig);
+            }
+
+            if (inputBuffer.Length == 0)
+            {
+                Debug.Assert(handshakeStarted == false);
+
+                handshakeStarted = true;
+                bool add = false;
+                // We grab reference to prevent disposal while handshake is pending.
+                this.DangerousAddRef(ref add);
+                ObjectDisposedException.ThrowIf(_disposed, this);
+                Interop.AppleCrypto.NwStartHandshake(SslContext, GCHandle.ToIntPtr(gcHandle));
```

Comment:
Seems like you might need some GC.KeepAlives in various places. This gcHandle is a weak handle, which means it's not keeping `this` alive, which means if nothing after the call to this instance method is guaranteed to be keeping the instance alive, it's possible it could be collected at a bad time around here.